### PR TITLE
Snow scuttling integration

### PR DESCRIPTION
**Pull request by weizman**
Use Snow to scuttle all future realms
* update `core` deps to install snow (also ignore snow on depcheck since it isn't consumed programatically)
* export snow source code from `core` package
* update `lavapack` to inject snow source code at the beginning of `runtime.js` if option `includeSnow` is `true`
* update `kernelTemplate` and `kernelCoreTemplate` to define snow (falling back to a dummy callback if snow intentionally is left out)
* update `kernelCoreTemplate` to use snow for scuttling
* regenerate `runtime.js`

### CODE REVIEW

File: packages/core/src/kernelTemplate.js [hunk 1]
```diff
@@ -1,5 +1,18 @@
 // LavaMoat Prelude
 (function () {
+  // identify the globalRef
+  const globalRef = (typeof globalThis !== 'undefined') ? globalThis : (typeof self !== 'undefined') ? self : (typeof global !== 'undefined') ? global : undefined
+  if (!globalRef) {
+    throw new Error('Lavamoat - unable to identify globalRef')
+  }
+
+  // polyfill globalThis
```
Comment (naugtur) [inline packages/core/src/kernelTemplate.js]: why is this necessary?

Comment (weizman) [reply packages/core/src/kernelTemplate.js]: don't know, that part isn't new though, i just moved it from a few lines later to here because I needed the determined `globalRef` object to get SNOW from it, and since that happenes in the outer scope I had to move it upwards a little bit.

File: packages/core/src/kernelCoreTemplate.js [hunk 2]
```diff
@@ -73,6 +73,7 @@
     const { getEndowmentsForConfig, makeMinimalViewOfRef, applyEndowmentPropDescTransforms } = templateRequire('makeGetEndowmentsForConfig')(generalUtils)
     const { prepareCompartmentGlobalFromConfig } = templateRequire('makePrepareRealmGlobalFromConfig')(generalUtils)
     const { strictScopeTerminator } = templateRequire('strict-scope-terminator')
+    const SNOW = globalRef.SNOW || ((cb, win) => cb(win))
```
Comment (kumavis) [inline packages/core/src/kernelCoreTemplate.js]: this fallback seems like its a "fail to unsafe" situation which is usually not what we want

File: packages/core/src/kernelTemplate.js (hunk 1 above)
Comment (kumavis) [reply packages/core/src/kernelTemplate.js]: `globalThis` was not common early in lavamoat's history. it may be prevalent enough to remove

**Comment by kumavis** [comment]
we kinda decided that we should not squash commits in a monorepo but that means having more meaningful commit messages

File: packages/browserify/test/fixtures/secureBundling/lavamoat/node/policy.json [hunk 3]
```diff
@@ -908,4 +908,4 @@
       }
     }
   }
-}
+}
```
Comment (legobeat) [inline packages/browserify/test/fixtures/secureBundling/lavamoat/node/policy.json]: This change should probably be removed?

File: packages/core/src/kernelTemplate.js (hunk 1 above)
Comment (weizman) [reply packages/core/src/kernelTemplate.js]: afba9f65d

File: packages/browserify/test/fixtures/secureBundling/lavamoat/node/policy.json (hunk 3 above)
Comment (weizman) [reply packages/browserify/test/fixtures/secureBundling/lavamoat/node/policy.json]: b4e0701fb

File: packages/core/src/kernelCoreTemplate.js (hunk 2 above)
Comment (weizman) [reply packages/core/src/kernelCoreTemplate.js]: 21a7787a6

File: packages/core/src/kernelTemplate.js (hunk 1 above)
Comment (legobeat) [reply packages/core/src/kernelTemplate.js]: > [afba9f6](https://github.com/LavaMoat/LavaMoat/commit/afba9f65d3b4e7404b5b374198c10838ef841458)

Seems like this belongs as a separate change (unless there's a link I'm missing)?

I propose soft-deprecating it for now and fully removing it in a subsequent release.

https://github.com/LavaMoat/LavaMoat/pull/459

File: packages/core/src/kernelTemplate.js [hunk 4]
```diff
@@ -1,5 +1,16 @@
...
 (function () {
+  const globalRef = globalThis;
+
+  const {
+    includeSnow,
+  } = __lavamoatBuilderOptions__
+
+  const SNOW = globalRef.SNOW;
+  if (!SNOW && includeSnow) {
+    throw 'LavaMoat - Snow is expected to exist but it does not'
```
Comment (legobeat) [inline packages/core/src/kernelTemplate.js:11]:
```suggestion
    throw new Error('LavaMoat - Snow is expected to exist but it does not');
```
?
Or if there's a motivation behind throwing a string, an explanatory comment would be great

File: packages/lavapack/src/runtime.js [hunk 5]
```diff
@@ -70,6 +70,17 @@
...
 (function () {
+  const globalRef = globalThis;
+
+  const {
+    includeSnow,
+  } = {"includeSnow":false}
+
+  const SNOW = globalRef.SNOW;
+  if (!SNOW && includeSnow) {
+    throw 'LavaMoat - Snow is expected to exist but it does not'
```
Comment (legobeat) [inline packages/lavapack/src/runtime.js:81]:
```suggestion
    throw new Error('LavaMoat - Snow is expected to exist but it does not');
```
?
Or if there's a motivation behind throwing a string, an explanatory comment would be great

File: packages/lavapack/src/runtime.js [hunk 6]
```diff
@@ -11142,6 +11142,11 @@ module.exports = {
   return module.exports
 })()
 
+    const SNOW = globalRef.SNOW;
+    if (!SNOW && includeSnow) {
+      throw 'LavaMoat - Snow is expected to exist but it does not';
```
Comment (legobeat) [inline packages/lavapack/src/runtime.js:11147]:
```suggestion
      throw new Error('LavaMoat - Snow is expected to exist but it does not');
```
?
Or if there's a motivation behind throwing a string, an explanatory comment would be great

File: packages/lavapack/src/runtime.js [hunk 7]
```diff
@@ -70,6 +70,17 @@
...
   const createKernel = // LavaMoat Prelude
 (function () {
+  const globalRef = globalThis;
+
+  const {
+    includeSnow,
+  } = {"includeSnow":false}
+
+  const SNOW = globalRef.SNOW;
+  if (!SNOW && includeSnow) {
```
Comment (legobeat) [inline packages/lavapack/src/runtime.js:80]: The `includeSnow` constant seems redundant...? (same applies to the other places with the same pattern)

```suggestion
  const { SNOW } = globalRef;
  if (!SNOW) {
```

Comment (legobeat) [inline packages/lavapack/src/runtime.js:80]: `includeSnow` is always `false` here right? Would it make sense to remove the no-op snow-specifics here in `runtime.js` until there is a way to set it picks up build-time configuration like the other entrypoints?

File: packages/lavapack/src/runtime.js [hunk 8]
```diff
@@ -70,6 +70,17 @@
 })({ onStatsReady }) : () => {}
   const createKernel = // LavaMoat Prelude
 (function () {
+  const globalRef = globalThis;
+
+  const {
+    includeSnow,
+  } = {"includeSnow":false}
+
```
Comment (legobeat) [inline packages/lavapack/src/runtime.js:78]:
```suggestion
  const includeSnow = false;
```
*(for clarity. in case this is to be kept in the first place for now, which is brought into question below)*

File: packages/core/src/kernelTemplate.js (hunk 4 above)
Comment (naugtur) [reply packages/core/src/kernelTemplate.js:11]: throwing a string means there's no stack trace. I'd recommend against it unless there's an important security reason.

Is this an attempt to tame errors by not using Errors?

File: packages/core/src/kernelTemplate.js [hunk 9]
```diff
@@ -22,21 +33,10 @@
...
-    const globalRef = (typeof globalThis !== 'undefined') ? globalThis : (typeof self !== 'undefined') ? self : (typeof global !== 'undefined') ? global : undefined
-    if (!globalRef) {
-      throw new Error('Lavamoat - unable to identify globalRef')
-    }
-
-    // polyfill globalThis
-    if (globalRef && !globalRef.globalThis) {
-      globalRef.globalThis = globalRef
-    }
-
```
Comment (naugtur) [inline packages/core/src/kernelTemplate.js:35]: I'm assuming this https://github.com/LavaMoat/LavaMoat/pull/459/files goes here as a result of merging/rebasing?

File: packages/lavapack/src/builder-runtime.js [hunk 10]
```diff
@@ -18,15 +18,20 @@ async function buildRuntimeCJS () {
...
+    output = getSnowSrc() + '\n\n' + output
+  }
   // inline kernel
   const kernelCode = generateKernel(opts)
   output = stringReplace(output, '__createKernel__', kernelCode)
   // inline reportStatsHook
   const statsCode = `(${makeInitStatsHook})({ onStatsReady })`
   output = stringReplace(output, '__reportStatsHook__', statsCode)
   output = markAsGenerated(output, 'runtime-template.js')
+  output = stringReplace(output, '__lavamoatBuilderOptions__', JSON.stringify({includeSnow}))
```
Comment (naugtur) [inline packages/lavapack/src/builder-runtime.js:34]: Why is this needed at runtime? If SNOW is enabled, should SNOW source not just run on its own, without needing extra config?

File: packages/lavapack/src/runtime-cjs.js [hunk 11]
```diff
@@ -2,16 +2,7 @@
...
-  // identify the globalRef
-  const globalRef = (typeof globalThis !== 'undefined') ? globalThis : (typeof self !== 'undefined') ? self : (typeof global !== 'undefined') ? global : undefined
-  if (!globalRef) {
-    throw new Error('Lavamoat - unable to identify globalRef')
-  }
-
-  // polyfill globalThis
-  if (globalRef && !globalRef.globalThis) {
-    globalRef.globalThis = globalRef
-  }
```
Comment (naugtur) [inline packages/lavapack/src/runtime-cjs.js:14]: @legobeat these files were not covered by your PR. Was that deliberate?
We probably can assume devs (who use this runtime) would not need a deprecation warning, but I want to make sure that's what you meant.

File: packages/core/src/kernelTemplate.js (hunk 9 above)
Comment (legobeat) [reply packages/core/src/kernelTemplate.js:35]: Indeed. The lavapack parts were not included there and are covered in #461.

File: packages/lavapack/src/runtime-cjs.js (hunk 11 above)
Comment (legobeat) [reply packages/lavapack/src/runtime-cjs.js:14]: It was indeed not deliberate! #461

**Comment by naugtur** [comment]
suggestion 
`rebase -i HEAD~15` before rebasing origin/master may save you time

**Comment by weizman** [comment]
decided on a different approach #462
